source_code/main_mcu/src/LOGIC/logic_totp.c

```diff
@@ -0,0 +1,174 @@
+                    // That means, we have shift_offset (which is also the number of remaining bits in the current byte) bits from the index remaining to put into the current byte.
+                    // Then we need to calculate the new shift_offset, which is basically the old shift_offset + 3 (As already described, the shift_offset is also
+                    // the number of remaining bits in the byte. As we have in the first step used this number of bits from the index, there are 5 - shift_offset bits from the
+                    // index remaining, that have to be put in the next byte. To get the number places the remaining bits have to be shifted in the next byte, we have to subtract
+                    // shift_offset from the number of bits in a byte. If we rearrange the order in that term, we get shift_offset = shift_offset + 8 - 5 = shift_offset + 3)
+                    // As an example for the second character, shift_offset is 3. So we take the first 3 of the 5 index bits and put them into current byte. Then we have 2 bits remaining.
+                    // Those bits need to be shifted then by 6 places to the left, to be at the beginning of the next byte.
+                    decoded_key[decoded_index] |= j >> (LOGIC_TOTP_BASE32_BITS - shift_offset);
+                    shift_offset += LOGIC_TOTP_BITS_PER_BYTE - LOGIC_TOTP_BASE32_BITS;
+                    decoded_index++;
+                    decoded_key[decoded_index] |= j << (shift_offset);
+                }
+            }
+        }
+    }
+}
+
+
+/*! \fn     logic_totp(char* totp_token, const char* key)
+*   \brief  Calculate TOTP Token using given key
+*   \param  calculated totp_token, char[6] expected
+*   \param  Base32 encoded key that is used to calculate totp_token, char[16] expectd
+*   \return void
+*/
+void logic_totp(char* totp_token, const char* key)
+{
+    char hash[20];
+    char decoded_key[10];
+    int truncated_hash = 0;
+    int offset = 0;
+    int shift_offset = LOGIC_TOTP_BITS_PER_BYTE * 3;
+    int token_idx = LOGIC_TOTP_TOKEN_LEN - 1;
+    time_t timestamp = logic_totp_get_time();
+    // step is the message the hash gets calculated for. TOTP is based on the current unix timestamp. As a TOTP token is valid for 30 seconds, we divide the current timestamp by
+    // 30 before calculating the SHA1 HMAC hash
+    time_t step = utils_switch_endiannes(timestamp/30);
+
+    logic_totp_base32_decode_key(decoded_key, key);
+    logic_totp_sha1_hmac(hash, decoded_key, &step);
+
+    // We use the last nibble of the calculated hash to get the offset. This is the offset in the hash, from where we take 4 bytes as truncated hash.
+    offset = hash[19] & 0x0F;
+
+    for (int i = offset; i < offset + LOGIC_TOTP_TRUNCATED_HASH_LEN; i++)
+    {
+        // Get the truncated hash as int, as we those 4 bytes as a whole number.
+        truncated_hash |= (hash[i] << shift_offset) & (0x000000FF << shift_offset);
```

Comment:
i'd go for the implementation from the RFC: https://tools.ietf.org/html/rfc6238
If you look at the page, it seems that (somehow?) the first bitshift is masked with 0x7F and **not** 0xFF

Reply:
yes, agree. just copy the code from there. That is what I did...